### DISCUSSION TITLE
Cube not illuminated by a light source

### DISCUSSION THREAD

**Discussion by Daoudou** [openscenegraph/OpenSceneGraph#1252] [Q&A] [answered]
Hello

I have a small problem with my light source.

I have 2 cubes, one built with the osg::Box class (the green cube) and the other built by hand (the red cube).

![85](https://github.com/openscenegraph/OpenSceneGraph/assets/98310378/9f6a61c5-d7bb-4b29-bf13-b6bdd415ca29)

As you can see only the underside of the green cube is illuminated by my light source but not the handmade red cube.

This may be due to the normal of my red cube? I am open to any explanation

the code of the red cube:

`     osg::ref_ptr<osg::Vec3Array> vertices = new osg::Vec3Array(8);
     (*vertices)[0].set(0.0f, 0.0f, 0.0f);
     (*vertices)[1].set(10.f, 0.f, 0.0f);
     (*vertices)[2].set(10.f, 0.f, 10.f);
     (*vertices)[3].set(0.f, 0.f, 10.0f);
     (*vertices)[4].set(10.f, 10.f, 0.0f);
     (*vertices)[5].set(0.0f, 10.0f, 0.0f);
     (*vertices)[6].set(0.0f, 10.0f, 10.0f);
     (*vertices)[7].set(10.0f, 10.0f, 10.0f);
 
     osg::ref_ptr<osg::DrawElementsUInt> indices =
         new osg::DrawElementsUInt(GL_TRIANGLES, 36);
     (*indices)[0] = 0; (*indices)[1] = 1; (*indices)[2] = 2;
     (*indices)[3] = 0; (*indices)[4] = 3; (*indices)[5] = 2;
     (*indices)[6] = 1; (*indices)[7] = 4; (*indices)[8] = 7;
     (*indices)[9] = 1; (*indices)[10] = 2; (*indices)[11] = 7;
     (*indices)[12] = 0; (*indices)[13] = 5; (*indices)[14] = 4;
     (*indices)[15] = 0; (*indices)[16] = 1; (*indices)[17] = 4;
     (*indices)[18] = 3; (*indices)[19] = 6; (*indices)[20] = 2;
     (*indices)[21] = 2; (*indices)[22] = 6; (*indices)[23] = 7;
     (*indices)[24] = 0; (*indices)[25] = 3; (*indices)[26] = 5;
     (*indices)[27] = 3; (*indices)[28] = 6; (*indices)[29] = 5;
     (*indices)[30] = 6; (*indices)[31] = 7; (*indices)[32] = 5;
     (*indices)[33] = 5; (*indices)[34] = 4; (*indices)[35] = 4;
 
     osg::ref_ptr<osg::Vec4Array> colors = new osg::Vec4Array;
     colors->push_back(osg::Vec4(1.0f, 0.0f, 0.0f, 1.0f));
     colors->push_back(osg::Vec4(1.0f, 0.0f, 0.0f, 1.0f));
     colors->push_back(osg::Vec4(1.0f, 0.0f, 0.0f, 1.0f));
     colors->push_back(osg::Vec4(1.0f, 0.0f, 0.0f, 1.0f));
 
     osg::ref_ptr<osg::Vec3Array> normals = new osg::Vec3Array;
     normals->push_back(osg::Vec3(0.0f, -1.0f, 0.0f));
 
     osg::ref_ptr<osg::Geometry> geom = new osg::Geometry;
     geom->setVertexArray(vertices.get());
     geom->addPrimitiveSet(indices.get());
     geom->setNormalArray(normals.get());
     geom->setNormalBinding(osg::Geometry::BIND_OVERALL);
     geom->setColorArray(colors.get());
     geom->setColorBinding(osg::Geometry::BIND_OVERALL);`

and the code for my light source:

` osg::ref_ptr<osg::Light> light = new osg::Light;
     light->setLightNum(0);
 
     osg::LightSource* lightSource = new osg::LightSource;
 
     light = lightSource->getLight();
     light->setLightNum(0);
     light->setAmbient(osg::Vec4d(0.52, 0.52, 0.52, 1.0f));
     light->setDiffuse(osg::Vec4d(0.72, 0.72, 0.72, 1.0));
     light->setSpecular(osg::Vec4d(0.30, 0.30, 0.30, 1.0f));
 
     osg::Vec3f direction(osg::Vec3f(0., 0., -20.));
     direction.normalize();
     light->setPosition(osg::Vec4f(direction, 0.0));  // directional light`

Thanks in advance :)

Daoudou

**Comment by robertosfield** [comment, marked as answer]
You can't light a cube with a single normal, it'll just light from single direction.

You'll need to use per vertex normals, and duplication the vertices for each face to enable you to set up the normals.  This is a bit awkward but it's down to the way that modern (last 20 years :-) now manage vertex shaders, even GL fixed function pipeline maps to this.
